ballet: add ChaCha20Rng

AlberErre opened the pull request:
This PR adds a `ChaCha20Rng` module to generate random number by using ChaCha20 algorithm 

It uses `openssl/evp.h` to implement ChaCha20

A test case (`test_fd_chacha20_rng.c`) has been added to verify its functionality, as well as a `README` explaining how to reproduce this on a MacOS machine

ptaffet-jump commented on src/ballet/chacha20/test_fd_chacha20_rng.c:
```diff
@@ -0,0 +1,60 @@
+#include <stdio.h>
```
If you need file IO, it should be protected with `#if FD_HAS_HOSTED`, but I don't think this test actually needs file IO.

AlberErre replied:
Just removed `stdio.h` as no  IO is needed

ptaffet-jump commented on src/ballet/chacha20/test_fd_chacha20_rng.c:
```diff
@@ -0,0 +1,60 @@
+/* Test fd_chacha20_rng_init function */
+void test_fd_chacha20_rng_init()
+{
+  unsigned char key[TEST_KEY_SIZE];
```
I don't know if we have a public comprehensive style guide yet, but these should be `uchar`

AlberErre replied:
Looks like `uchar` type is not directly available In OpenSSL

I'm getting the following error while trying to use it -> `error: unknown type name 'uchar'; did you mean 'char'?`

AlberErre replied:
thanks! just solved this by adding `#include "../fd_ballet_base.h"` , which uses `fd_util.h` under the hood 👍🏻

ptaffet-jump commented on src/ballet/chacha20/test_fd_chacha20_rng.c:
```diff
@@ -0,0 +1,60 @@
+  unsigned char nonce[TEST_NONCE_SIZE];
+
+  /* Generate key and nonce using /dev/urandom */
+  FILE *fp = fopen("/dev/urandom", "rb");
+  fread(key, sizeof(key), 1, fp);
+  fread(nonce, sizeof(nonce), 1, fp);
+  fclose(fp);
```
Why add a dependency to urandom when you don't actually care about the randomness? I'd suggest initializing it to something you know (like all 1s) or using `fd_rng` with a fixed seed.

AlberErre replied:
totally agree, just removed it ✅ using a test vector instead

ptaffet-jump commented on src/ballet/chacha20/test_fd_chacha20_rng.c:
```diff
@@ -0,0 +1,60 @@
+  fclose(fp);
+
+  /* Initialize the random number generator */
+  assert(fd_chacha20_rng_init(key, nonce) == 0);
```
We normally prefer `FD_TEST` over `assert` for a unit test

AlberErre replied:
updated ✅

ptaffet-jump commented on src/ballet/chacha20/test_fd_chacha20_rng.c:
```diff
@@ -0,0 +1,60 @@
+  assert(fd_chacha20_rng_init(key, nonce) == 0);
+
+  /* Generate two random numbers and make sure they are different */
+  assert(fd_chacha20_rng_get_uint32(&num1) == 0);
+  assert(fd_chacha20_rng_get_uint32(&num2) == 0);
+  assert(num1 != num2);
```
We need to make this test much stronger than just checking that the two numbers are different. Richie suggested adding test vectors. I think that's a good idea. You could also add some light randomness tests

AlberErre replied:
just added a test vector ✅

ptaffet-jump commented on src/ballet/chacha20/fd_chacha20_rng.c:
```diff
@@ -0,0 +1,33 @@
+#include "fd_chacha20.h"
+
+static EVP_CIPHER_CTX *ctx = NULL;
```
Firedancer doesn't have much static state like this. Prefer the footprint/align/alloc/new pattern that the bulk of the code follows. If that doesn't work in this situation, this needs a big comment block explaining why.

AlberErre replied:
totally agree, deleted this static state ✅, now we generates this EVP_CIPHER_CTX per call and free it from memory once it has finished

ptaffet-jump commented on src/ballet/chacha20/fd_chacha20.h:
```diff
@@ -0,0 +1,14 @@
+#define FD_CHACHA20_KEY_SIZE 32   // 32 bytes, 256 bits
+#define FD_CHACHA20_NONCE_SIZE 12 // 12 bytes, 96 bits
+
+int fd_chacha20_rng_init(unsigned char *key, unsigned char *nonce);
+int fd_chacha20_rng_get_uint32(fd_chacha20_rng_t *num);
```
These functions in the header need some documentation. See other header files in the project to get a sense of what I mean.

AlberErre replied:
Just  updated this by deleting these methods from header (there are now internal functions). Now we're exposing a single method to generate random numbers

AlberErre replied:
Definitely need to add some comments to document this as well

ptaffet-jump commented on src/ballet/chacha20/README.md:
````diff
@@ -0,0 +1,15 @@
+Build binary by using the following command
+
+```
+gcc -o test_fd_chacha20_rng test_fd_chacha20_rng.c fd_chacha20_rng.c -I/usr/local/opt/openssl/include -L/usr/local/opt/openssl/lib -lcrypto
+```
````
Adding a README is nice, but the normal way to build a test is with `make`. Does this build with `make`? If so, you should find the unit test in the unit-tests directory.

ptaffet-jump replied:
The unit test should also probably be called `test_chacha20` without the `_rng` suffix.

AlberErre replied:
All renaming will be done once code from `openssl` is imported.

Regarding the `make` file, I need to get my feets wet on that 👀

AlberErre replied:
just added a Makefile and updated the README docs ✅

ptaffet-jump commented on src/ballet/chacha20/fd_chacha20_rng.c:
I think this should be renamed to `fd_chacha20.c`

AlberErre replied:
ok! will rename it once code from openssl is imported

AlberErre replied:
renamed ✅

ptaffet-jump commented on src/ballet/chacha20/test_fd_chacha20_rng.c:
```diff
@@ -0,0 +1,60 @@
+  test_fd_chacha20_rng_init();
+  test_fd_chacha20_rng_get_uint32();
+
+  printf("All fd_chacha20_rng tests passed!\n");
```
This should be FD_LOG_NOTICE

AlberErre replied:
updated ✅

ptaffet-jump reviewed (changes requested):
Hi there, thanks for the contribution! There are a few places where your style doesn't match the firedancer code base that definitely need to get fixed before this is merged. Mostly, I'd echo Richie: we would rather copy a few files of code from openssl (so that we can standardize and optimize it) than call out to the library. The other significant issue is the use of the static pointer. That seems to tie our hands in a way we might regret in the future.

AlberErre commented:
I've refactored the implementation to expose a single method `fd_chacha20_generate_random_number` as the API.

Also, I've updated some stuff based on your comments ✅

Next step will be to copy source from `openssl` directly, as suggested by @ripatel-jump  

I'll keep you posted on discord in case I have any questions 🙏🏼

AlberErre commented:
@ripatel-jump @ptaffet-jump great news! 🎉 

I finally was able to reproduce the same random number as the rust crate, by using a C implementation ✅ (needed to play with nonce and counter params to achieve it).

I've updated the test cases & `Makefile` accordingly 👍🏻 - Now, test cases directly compare the output from both C implementation and FFI rust crate functions, to secure both implementations produces the same random number deterministically 🎉

Next, I will copy the OpenSSL source code as you pointed out @ripatel-jump. Totally understand your point about porting Solana's cryptographic algorithms to C.

AlberErre commented:
Great! 🎉 I'm happy to be able to help with this 🙏🏼 just learned a lot about ciphers
